File: test/devices/garmin/garmin_notification_actions_test.dart
import 'dart:typed_data';

import 'package:flutter_test/flutter_test.dart';
import 'package:notification_listener_native/notification_listener_native.dart';

import 'package:openvitals/devices/garmin/garmin_notification_actions.dart';
import 'package:openvitals/devices/garmin/garmin_notification_messages.dart';

NotificationMsg _message({
  List<NotificationActionMsg> actions = const [],
  bool dismissable = true,
}) =>
    NotificationMsg(
      id: 7,
      packageName: 'com.example.chat',
      appLabel: 'Chat',
      title: 'Ada',
      body: 'On my way',
      whenEpochMillis: 0,
      categoryOrdinal: GarminNotificationCategory.sms.index,
      removed: false,
      actions: actions,
      dismissable: dismissable,
    );

NotificationActionMsg _action(
  int index,
  String title, {
  bool reply = false,
  bool fireable = true,
}) =>
    NotificationActionMsg(
      index: index,
      title: title,
      isReply: reply,
      fireableFromBackground: fireable,
    );

void main() {
  group('mapping Android actions onto the watch', () {
    test('every dismissable notification gets a dismiss the app did not '
        'provide', () {
      // Android has no "clear this" action — clearing is the listener's job —
      // so it has to be synthesised, and marked so the performer cancels
      // instead of firing an intent that does not exist.
      final actions = garminActionsFor(_message());

      expect(actions.single.kind, GarminNotificationActionKind.dismiss);
      expect(actions.single.isSynthetic, isTrue);
      expect(actions.single.androidIndex, -1);
    });

    test('an ongoing notification gets no dismiss, because clearing it would '
        'fail', () {
      expect(garminActionsFor(_message(dismissable: false)), isEmpty);
    });

    test("the app's own buttons land in the numbered custom slots, in order",
        () {
      final actions = garminActionsFor(_message(actions: [
        _action(0, 'Mark as read'),
        _action(1, 'Snooze'),
      ]));

      expect(actions.map((a) => a.kind), [
        GarminNotificationActionKind.dismiss,
        GarminNotificationActionKind.custom1,
        GarminNotificationActionKind.custom2,
      ]);
      expect(actions[1].label, 'Mark as read');
      expect(actions[1].androidIndex, 0);
    });

    test('a reply action takes the reply slot, not a custom one', () {
      final actions = garminActionsFor(_message(actions: [
        _action(0, 'Reply', reply: true),
        _action(1, 'Mark as read'),
      ]));

      final reply = actions
          .firstWhere((a) => a.kind == GarminNotificationActionKind.reply);
      expect(reply.isReply, isTrue);
      expect(reply.androidIndex, 0);
      expect(
        actions.any((a) => a.kind == GarminNotificationActionKind.custom1),
        isTrue,
      );
    });

    test('a second reply becomes a plain button rather than overwriting the '
        'first', () {
      // The watch has one reply control, so a second would silently replace it.
      final actions = garminActionsFor(_message(actions: [
        _action(0, 'Reply', reply: true),
        _action(1, 'Reply privately', reply: true),
      ]));

      expect(
        actions.where((a) => a.kind == GarminNotificationActionKind.reply),
        hasLength(1),
      );
      final second = actions.firstWhere((a) => a.androidIndex == 1);
      expect(second.kind, GarminNotificationActionKind.custom1);
      expect(second.isReply, isFalse);
    });

    test('more buttons than there are slots are dropped, not crammed in', () {
      // Reusing a slot would make two controls invoke the same thing.
      final actions = garminActionsFor(_message(actions: [
        for (var i = 0; i < 8; i++) _action(i, 'Action $i'),
      ]));

      final customs = actions
          .where((a) => GarminNotificationActionKind.customSlots
              .contains(a.kind))
          .toList();
      expect(customs, hasLength(maxGarminCustomActions));
      expect(customs.map((a) => a.androidIndex), [0, 1, 2, 3, 4]);
    });

    test('an action that only opens the app is not offered at all', () {
      // A stock SMS app publishes a "Reply" that prefills its compose screen
      // rather than sending, and Android blocks a background activity launch —
      // so it does nothing, silently. Offering it puts back the dead button
      // this feature exists to remove.
      final actions = garminActionsFor(_message(actions: [
        _action(0, 'Reply', reply: true, fireable: false),
        _action(1, 'Mark as read'),
      ]));

      expect(
        actions.any((a) => a.kind == GarminNotificationActionKind.reply),
        isFalse,
      );
      expect(actions.map((a) => a.label), contains('Mark as read'));
    });

    test('a blocked reply does not consume the reply slot, so a later usable '
        'one still gets it', () {
      final actions = garminActionsFor(_message(actions: [
        _action(0, 'Reply', reply: true, fireable: false),
        _action(1, 'Quick reply', reply: true),
      ]));

      final reply = actions
          .firstWhere((a) => a.kind == GarminNotificationActionKind.reply);
      expect(reply.androidIndex, 1);
    });

    test('an index survives the round trip, so the phone never re-derives which '
        'button was meant', () {
      final actions = garminActionsFor(_message(actions: [
        _action(3, 'Archive'),
      ]));
      expect(actions.last.androidIndex, 3);
    });
  });

  group('encoding the ACTIONS attribute', () {
    test('no actions encodes as the four-zero-byte sentinel', () {
      expect(
        encodeGarminNotificationActions(const []),
        Uint8List.fromList(const [0, 0, 0, 0]),
      );
    });

    test('each action is a code, an icon position, a length and a label', () {
      final bytes = encodeGarminNotificationActions(const [
        GarminNotificationAction(
          kind: GarminNotificationActionKind.custom1,
          label: 'Ok',
          androidIndex: 0,
        ),
      ]);

      expect(
        bytes,
        Uint8List.fromList(const [
          1, // one action
          1, // CUSTOM_ACTION_1
          0, // no fixed icon position
          2, // label length
          0x4F, 0x6B, // "Ok"
        ]),
      );
    });

    test('dismiss carries the LEFT icon position, which is where the watch '
        'draws it', () {
      final bytes = encodeGarminNotificationActions(const [
        GarminNotificationAction(
          kind: GarminNotificationActionKind.dismiss,
          label: 'X',
          androidIndex: -1,
        ),
      ]);
      expect(bytes[1], 98); // DISMISS_NOTIFICATION
      expect(bytes[2], GarminActionIconPosition.left.bit);
    });

    test('reply carries the BOTTOM icon position', () {
      final bytes = encodeGarminNotificationActions(const [
        GarminNotificationAction(
          kind: GarminNotificationActionKind.reply,
          label: 'R',
          androidIndex: 0,
          isReply: true,
        ),
      ]);
      expect(bytes[1], 95); // REPLY_MESSAGES
      expect(bytes[2], GarminActionIconPosition.bottom.bit);
    });

    test('the label length is BYTES, so a non-ASCII label still parses', () {
      final bytes = encodeGarminNotificationActions(const [
        GarminNotificationAction(
          kind: GarminNotificationActionKind.custom1,
          label: 'áé',
          androidIndex: 0,
        ),
      ]);
      expect(bytes[3], 4, reason: 'two characters, four UTF-8 bytes');
    });

    test('an absurdly long label is trimmed rather than wrapping the length '
        'byte', () {
      // A length byte cannot carry more than 255, and wrapping would make the
      // watch read the next action code as label text.
      final bytes = encodeGarminNotificationActions([
        GarminNotificationAction(
          kind: GarminNotificationActionKind.custom1,
          label: 'x' * 400,
          androidIndex: 0,
        ),
      ]);
      expect(bytes[3], 255);
      expect(bytes.length, 4 + 255);
    });

    test('several actions pack one after another', () {
      final bytes = encodeGarminNotificationActions(const [
        GarminNotificationAction(
          kind: GarminNotificationActionKind.dismiss,
          label: 'a',
          androidIndex: -1,
        ),
        GarminNotificationAction(
          kind: GarminNotificationActionKind.custom1,
          label: 'bb',
          androidIndex: 0,
        ),
      ]);
      expect(bytes[0], 2);
      // count + (3 + 1) + (3 + 2)
      expect(bytes.length, 1 + 4 + 5);
    });
  });
}

File: app/src/test/kotlin/tech/mmarca/openvitals/devices/garmin/GarminNotificationActionsTest.kt
package tech.mmarca.openvitals.devices.garmin

import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test
import tech.mmarca.openvitals.devices.notifications.NotificationActionMsg
import tech.mmarca.openvitals.devices.notifications.NotificationMsg

/** Mapping an Android notification's actions onto GNCS's fixed slots, and encoding the ACTIONS attribute. */
class GarminNotificationActionsTest {

    private fun message(
        actions: List<NotificationActionMsg> = emptyList(),
        dismissable: Boolean = true,
    ) = NotificationMsg(
        id = 7,
        packageName = "com.example.chat",
        appLabel = "Chat",
        title = "Ada",
        body = "On my way",
        whenEpochMillis = 0,
        categoryOrdinal = GarminNotificationCategory.SMS.ordinal,
        removed = false,
        actions = actions,
        dismissable = dismissable,
    )

    private fun action(
        index: Int,
        title: String,
        reply: Boolean = false,
        fireable: Boolean = true,
    ) = NotificationActionMsg(
        index = index,
        title = title,
        isReply = reply,
        fireableFromBackground = fireable,
    )

    // Mapping Android actions onto the watch.

    @Test
    fun `every dismissable notification gets a dismiss the app did not provide`() {
        // Android has no "clear this" action, so it is synthesised and marked so the performer cancels.
        val actions = garminActionsFor(message())

        assertEquals(GarminNotificationActionKind.DISMISS, actions.single().kind)
        assertTrue(actions.single().isSynthetic)
        assertEquals(-1, actions.single().androidIndex)
    }

    @Test
    fun `an ongoing notification gets no dismiss, because clearing it would fail`() {
        assertTrue(garminActionsFor(message(dismissable = false)).isEmpty())
    }

    @Test
    fun `the app's own buttons land in the numbered custom slots, in order`() {
        val actions = garminActionsFor(
            message(
                actions = listOf(
                    action(0, "Mark as read"),
                    action(1, "Snooze"),
                ),
            ),
        )

        assertEquals(
            listOf(
                GarminNotificationActionKind.DISMISS,
                GarminNotificationActionKind.CUSTOM_1,
                GarminNotificationActionKind.CUSTOM_2,
            ),
            actions.map { it.kind },
        )
        assertEquals("Mark as read", actions[1].label)
        assertEquals(0, actions[1].androidIndex)
    }

    @Test
    fun `a reply action takes the reply slot, not a custom one`() {
        val actions = garminActionsFor(
            message(
                actions = listOf(
                    action(0, "Reply", reply = true),
                    action(1, "Mark as read"),
                ),
            ),
        )

        val reply = actions.first { it.kind == GarminNotificationActionKind.REPLY }
        assertTrue(reply.isReply)
        assertEquals(0, reply.androidIndex)
        assertTrue(actions.any { it.kind == GarminNotificationActionKind.CUSTOM_1 })
    }

    @Test
    fun `a second reply becomes a plain button rather than overwriting the first`() {
        // The watch has one reply control, so a second would replace it.
        val actions = garminActionsFor(
            message(
                actions = listOf(
                    action(0, "Reply", reply = true),
                    action(1, "Reply privately", reply = true),
                ),
            ),
        )

        assertEquals(1, actions.count { it.kind == GarminNotificationActionKind.REPLY })
        val second = actions.first { it.androidIndex == 1 }
        assertEquals(GarminNotificationActionKind.CUSTOM_1, second.kind)
        assertFalse(second.isReply)
    }

    @Test
    fun `more buttons than there are slots are dropped, not crammed in`() {
        // Reusing a slot would make two controls invoke the same thing.
        val actions = garminActionsFor(
            message(actions = (0 until 8).map { action(it, "Action $it") }),
        )

        val customs = actions.filter { it.kind in GarminNotificationActionKind.customSlots }
        assertEquals(MAX_GARMIN_CUSTOM_ACTIONS, customs.size)
        assertEquals(listOf(0, 1, 2, 3, 4), customs.map { it.androidIndex })
    }

    @Test
    fun `an action that only opens the app is not offered at all`() {
        // A stock SMS "Reply" prefills a compose screen, which Android blocks in the background. Offering it puts back a dead button.
        val actions = garminActionsFor(
            message(
                actions = listOf(
                    action(0, "Reply", reply = true, fireable = false),
                    action(1, "Mark as read"),
                ),
            ),
        )

        assertFalse(actions.any { it.kind == GarminNotificationActionKind.REPLY })
        assertTrue(actions.map { it.label }.contains("Mark as read"))
    }

    @Test
    fun `a blocked reply does not consume the reply slot, so a later usable one still gets it`() {
        val actions = garminActionsFor(
            message(
                actions = listOf(
                    action(0, "Reply", reply = true, fireable = false),
                    action(1, "Quick reply", reply = true),
                ),
            ),
        )

        val reply = actions.first { it.kind == GarminNotificationActionKind.REPLY }
        assertEquals(1, reply.androidIndex)
    }

    @Test
    fun `an index survives the round trip, so the phone never re-derives which button was meant`() {
        val actions = garminActionsFor(message(actions = listOf(action(3, "Archive"))))
        assertEquals(3, actions.last().androidIndex)
    }

    // Encoding the ACTIONS attribute.

    @Test
    fun `no actions encodes as the four-zero-byte sentinel`() {
        assertArrayEquals(
            byteArrayOf(0, 0, 0, 0),
            encodeGarminNotificationActions(emptyList()),
        )
    }

    @Test
    fun `each action is a code, an icon position, a length and a label`() {
        val bytes = encodeGarminNotificationActions(
            listOf(
                GarminNotificationAction(
                    kind = GarminNotificationActionKind.CUSTOM_1,
                    label = "Ok",
                    androidIndex = 0,
                ),
            ),
        )

        assertArrayEquals(
            byteArrayOf(
                1, // one action
                1, // CUSTOM_ACTION_1
                0, // no fixed icon position
                2, // label length
                0x4F, 0x6B, // "Ok"
            ),
            bytes,
        )
    }

    @Test
    fun `dismiss carries the LEFT icon position, which is where the watch draws it`() {
        val bytes = encodeGarminNotificationActions(
            listOf(
                GarminNotificationAction(
                    kind = GarminNotificationActionKind.DISMISS,
                    label = "X",
                    androidIndex = -1,
                ),
            ),
        )
        assertEquals(98, bytes[1].toInt() and 0xFF) // DISMISS_NOTIFICATION
        assertEquals(GarminActionIconPosition.LEFT.bit, bytes[2].toInt())
    }

    @Test
    fun `reply carries the BOTTOM icon position`() {
        val bytes = encodeGarminNotificationActions(
            listOf(
                GarminNotificationAction(
                    kind = GarminNotificationActionKind.REPLY,
                    label = "R",
                    androidIndex = 0,
                    isReply = true,
                ),
            ),
        )
        assertEquals(95, bytes[1].toInt() and 0xFF) // REPLY_MESSAGES
        assertEquals(GarminActionIconPosition.BOTTOM.bit, bytes[2].toInt())
    }

    @Test
    fun `the label length is BYTES, so a non-ASCII label still parses`() {
        val bytes = encodeGarminNotificationActions(
            listOf(
                GarminNotificationAction(
                    kind = GarminNotificationActionKind.CUSTOM_1,
                    label = "áé",
                    androidIndex = 0,
                ),
            ),
        )
        assertEquals("two characters, four UTF-8 bytes", 4, bytes[3].toInt())
    }

    @Test
    fun `an absurdly long label is trimmed rather than wrapping the length byte`() {
        // A length byte cannot carry more than 255; wrapping would read the next code as label text.
        val bytes = encodeGarminNotificationActions(
            listOf(
                GarminNotificationAction(
                    kind = GarminNotificationActionKind.CUSTOM_1,
                    label = "x".repeat(400),
                    androidIndex = 0,
                ),
            ),
        )
        assertEquals(255, bytes[3].toInt() and 0xFF)
        assertEquals(4 + 255, bytes.size)
    }

    @Test
    fun `several actions pack one after another`() {
        val bytes = encodeGarminNotificationActions(
            listOf(
                GarminNotificationAction(
                    kind = GarminNotificationActionKind.DISMISS,
                    label = "a",
                    androidIndex = -1,
                ),
                GarminNotificationAction(
                    kind = GarminNotificationActionKind.CUSTOM_1,
                    label = "bb",
                    androidIndex = 0,
                ),
            ),
        )
        assertEquals(2, bytes[0].toInt())
        // count + (3 + 1) + (3 + 2)
        assertEquals(1 + 4 + 5, bytes.size)
    }
}
